C23_WMDE_Desktop_DE_17

gbirke opened the pull request:
Ticket: https://phabricator.wikimedia.org/T352507

gbirke commented on pad_preact/components/ui/form/DonationForm.jsx:
```diff
@@ -0,0 +1,147 @@
+
+	return <div className="form">
+		<form method="post" name="donationForm" className="form__element"
+			action={ formAction }>
```
The campaign parameters (`props.bannerName` and `props.campaignName`) are missing.

gbirke commented on pad_preact/components/ui/form/FormItemsBuilder.js:
```diff
@@ -0,0 +1,59 @@
+	NO: { value: 'anonym', label: 'give-address-statement-negative' }
+};
+
+export default class FormItemsBuilder {
```
Since the other form items are 100% identical, please use inheritance here. This will make the class much shorter.

gbirke commented on shared/messages/de.js:
```diff
@@ -48,7 +48,10 @@ const Translations = {
+	'payment-sofort': 'Sofortüberweisung',
+
+	'give-address-statement-positive': 'Ja',
+	'give-address-statement-negative': 'Nein'
```
The translations file is part of every banner, not only the iPad banner. Therefore, we should keep a certain discipline to mark translations that are part of an A/B test as such, to potentially clean up after the tets is done.

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,117 @@
+				locale='de'/>
+		</div>;
+	}
+
```
Please remove the `FundModal` component and all state and state-changing methods from this banner. For iPad, the banner just displays a link.

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,117 @@
+		onClose: PropTypes.func,
+		/** */
+		registerDisplayBanner: PropTypes.func.isRequired
+	}
```
I'm not sure what's worse - an incomplete list of available properties or no list of properties at all. What do you think?

moiikana replied:
would a complete list of properties in this case be the following? (+their respective types) : 
- trackingIds: ,
- campaignParameters: ,
- campaignProjection: ,
- formatters: ,
- bannerText: ,
- donationForm: ,
- translations: ,
- formItems:
- trackingData: this.trackingData,
- impressionCounts:,
- onClose: PropTypes.func,
- registerDisplayBanner: PropTypes.func.isRequired

these are the properties from banner_ctrl.js, banner_presenter.js and Banner.jsx.
I think it would be nice to have them documented like this, if this list is complete and correct.

gbirke commented on pad_preact/Banner_var.jsx:
```diff
@@ -0,0 +1,117 @@
+		</div>;
+	}
+
+}
```
Please delete `Banner_var.jsx`, that's unneccessary duplication. Instead, load different form components and pass in the form component as a property (like we do for BannerText)

gbirke commented on shared/messages/de.js:
```diff
@@ -48,7 +48,11 @@ const Translations = {
-	'payment-sofort': 'Sofortüberweisung'
+	'payment-sofort': 'Sofortüberweisung',
+
+	// the following 2 lines are only for the ipad var test, remove if not in use anymore
```
👍

gbirke commented on pad_preact/styles/styles_ctrl.pcss:
```diff
@@ -0,0 +1,12 @@
+@import './Infobox.pcss';
+@import './BannerText.pcss';
+@import './ProgressBar.pcss';
+@import '../../shared/components/ui/FundsModal.pcss';
```
You can remove this line, as we don't have a FundsModal here

moiikana commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,118 @@
+export default class Banner extends Component {
+
+	static propTypes = {
+		campaignParameters: PropTypes.object,
```
I tried to accumulate the propTypes that are passed down to the Banner class by banner_ctrl.js, banner_presenter and so on..
Is it correct that they all should be defined here? (they are used here, so maybe yes..)
I tried to give them the most fitting type and tried to check my guessings with `console.log(typeof ....)` , but that always told me it's either an object or a func. I'm also not sure about whether it could be PropTypes.elementType or .element in line 23-24.

gbirke replied:
Thank you so much! 

Two props are missing: `bannerName` and `campaignName` (coming from `...trackingIds`) in the entry point.

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,118 @@
+		campaignProjection: PropTypes.instanceOf( CampaignProjection ), //object?
+		formatters: PropTypes.object,
+		bannerText: PropTypes.elementType, //func?
+		donationForm: PropTypes.elementType, //func?
```
`elementType` is exactly right. Elements can be either functions or classes.

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,118 @@
+		bannerText: PropTypes.elementType, //func?
+		donationForm: PropTypes.elementType, //func?
+		translations: PropTypes.object,
+		formItems: PropTypes.object,
```
```suggestion
                /**
                   *  Contains items with values and labels for SelectGroup components 
                   *  in the form, e.g. amounts, payment types, intervals, etc 
                   */
		formItems: PropTypes.object,
```

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,118 @@
+
+		/** callback when banner closes */
+		onClose: PropTypes.func,
+		registerDisplayBanner: PropTypes.func.isRequired
```
```suggestion
                /** Callback to register a displayBanner function with the BannerPresenter */
		registerDisplayBanner: PropTypes.func.isRequired
```

gbirke commented on pad_preact/Banner.jsx:
```diff
@@ -0,0 +1,118 @@
+		translations: PropTypes.object,
+		formItems: PropTypes.object,
+
+		/** complex object that holds any data needed for tracking purposes */
```
```suggestion
		/** object that holds any data needed for tracking purposes */
```

gbirke committed:
Prepare Campaign for C23_WMDE_Desktop_DE_17

Based on ctrl of C23_WMDE_Desktop_DE_16

Left styles_var in place so the Git diff shows the changes better.

Added tests for "Already Donated" modal

Ticket: https://phabricator.wikimedia.org/T352507

gbirke committed:
Implement VAR for C23_WMDE_Desktop_DE_17

New blue design

Ticket: https://phabricator.wikimedia.org/T352507

gbirke commented:
I based this on the Desktop 16 branch, will rebase & adapt this when that branch is merged.

If the VAR banner wins, we'll have to discuss if we create a new Theme (basically a clone of Treedip with different colors and some slider style changes) or how we structure theme variants to keep backwards compatibility

Abban commented:
To me this looks like it should be a new theme, based on the amount of overrides happening in the banner styles.

gbirke committed:
Archive C23_WMDE_Desktop_DE_16

Also fix import order of Desktop 16 to restore some sense to the
comments for the import sections

Ticket: https://phabricator.wikimedia.org/T352507

Abban commented:
Can you add the Already Donated feature tests to the banner tests? I forgot to add them when I added it back to the banner.

```js
describe( 'Already Donated', () => {
	test.each( [
		[ 'expectShowsAlreadyDonatedModal' ],
		[ 'expectHidesAlreadyDonatedModal' ],
		[ 'expectFiresMaybeLaterEvent' ],
		[ 'expectFiresGoAwayEvent' ]
	] )( '%s', async ( testName: string ) => {
		await alreadyDonatedModalFeatures[ testName ]( getWrapper() );
	} );
} );
```

Abban commented on test/banners/desktop/components/BannerVar.spec.ts:
```diff
@@ -1,6 +1,6 @@
+import { afterEach, beforeEach, describe, test, vi } from 'vitest';
 import { mount, VueWrapper } from '@vue/test-utils';
-import Banner from '../../../../banners/desktop/components/BannerVar.vue';
+import Banner from '../../../../banners/desktop/components/BannerCtrl.vue';
```
This should import BannerVar

Abban commented on test/banners/desktop/components/BannerVar.spec.ts:
```diff
@@ -19,12 +20,12 @@ import { useFormModel } from '@src/components/composables/useFormModel';
 const translator = ( key: string ): string => key;
 
-describe( 'BannerVar.vue', () => {
+describe( 'BannerCtrl.vue', () => {
```
This should be BannerVar

Abban reviewed (approved):
Looks good, there's just a little clean up to do before merging.

There are some files you should delete:

- `visitorsVsDailyDonorsSentence.ts`
- `styles/MainBanner_var.scss`

Your IDE is adding 8 space indents to HTML, you can fix that in PhpStorm by going to `Settings > Editor > Code Style > HTML` and making sure all indents are set to 4
